client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -399,17 +398,33 @@ function mediaButton( editor ) {
 		delete media.caption;
 		media = assign( {}, parsed.media, media );
 
+		// Compute a ratio used to order media sizes
+		const computeRatio = ( { width = media.width, height = media.height } ) => {
+			return Math.min(
+				( width / media.width ) || Infinity,
+				( height / media.height ) || Infinity
+			);
+		};
+
 		// Determine the next usable size
-		let sizeIndex = SIZE_ORDER.indexOf( parsed.appearance.size );
+		const sizeRatios = SIZE_ORDER
+			.map( size => computeRatio( MediaUtils.getThumbnailSizeDimensions( size, site ) ) );
+		const sizeIndex = SIZE_ORDER.indexOf( parsed.appearance.size );
+		const displayedRatio = sizeIndex !== -1 ? sizeRatios[ sizeIndex ] : computeRatio( parsed.media );
+		const possibleSizes = SIZE_ORDER
+			.filter( ( size, index ) => {
+				return sizeRatios[ index ] <= 1 && (
+					increment > 0
+						? sizeRatios[ index ] > displayedRatio
+						: sizeRatios[ index ] < displayedRatio
+					);
+			} );
 		let size;
-		do {
-			sizeIndex += increment;
-			size = SIZE_ORDER[ Math.max( sizeIndex, 0 ) ];
-			const dimensions = MediaUtils.getThumbnailSizeDimensions( size, site );
-			if ( media.width > dimensions.width || media.height > dimensions.width ) {
-				break;
-			}
-		} while ( sizeIndex > 0 && sizeIndex < SIZE_ORDER.length - 1 );
+		if ( possibleSizes.length ) {
+			size = increment > 0 ? possibleSizes[ 0 ] : possibleSizes[ possibleSizes.length - 1 ];
```

Comment:
I wonder if this would be a good candidate for Lodash [`find`](https://lodash.com/docs/4.16.4#find) and [`findLast`](https://lodash.com/docs/4.16.4#findLast). It would be more performant to stop at the first discovered match, and more meaningful (since we never use more than a single member of the filtered set).

Something like this?

```js
function isMatchingSize( size, index ) {
	if ( sizeRatios[ index ] > 1 ) {
		return false;
	}

	if ( increment > 0 ) {
		return sizeRatios[ index ] > displayedRatio;
	}


	return sizeRatios[ index ] < displayedRatio;
}

const findFn = increment > 0 ? find : findLast;
const size = findFn( SIZE_ORDER, isMatchingSize ) || SIZE_ORDER[ SIZE_ORDER.length - 1 ];
```